### PR TITLE
SQOOP-3419: Refactor S3 tests for easier extensibility

### PR DESCRIPTION
**Pull request by szvasas**
This patch covers the following changes:

- The original TestS3* tests are made abstract and generic, moved to the cloud package instead. The concrete cloud storage tests extend these abstract tests.
- S3TestUtils is broken into 3 classes: CloudCredentialsRule, CloudFileSystemRule and CloudTestDataSet
- S3CredentialGenerator and its only implementation DefaultS3CredentialGenerator have been simplified in the new CredentialGenerator class which is reusable for different cloud storage providers

### CODE REVIEW

File: src/test/org/apache/sqoop/cloud/CloudImportJobTestCase.java [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+
+  protected void createTestTableFromInitialInputDataForMerge() {
+    createTableWithRecordsWithColTypesAndNames(dataSet.getColumnNamesForMerge(), dataSet.getColumnTypesForMerge(), dataSet.getInitialInputDataForMerge());
+  }
+
+  protected void insertInputDataIntoTableForMerge(List<List<Object>> inputData) {
+    insertRecordsIntoTableWithColTypesAndNames(dataSet.getColumnNamesForMerge(), dataSet.getColumnTypesForMerge(), inputData);
+  }
+
+  protected void failIfOutputFilePathContainingPatternExists(FileSystem s3Client, Path targetDirPath, String pattern) throws IOException {
```
Comment (ebogi) [inline src/test/org/apache/sqoop/cloud/CloudImportJobTestCase.java]: Renaming of s3Client parameter has been missed. I would rename it to cloudFileSystem for example.

File: src/test/org/apache/sqoop/cloud/CloudImportJobTestCase.java [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    List<Path> outputFilesWithPathContainingPattern = FileSystemUtil.findFilesWithPathContainingPattern(targetDirPath,
+        s3Client.getConf(), pattern);
+    if (outputFilesWithPathContainingPattern.size() != 0) {
+      fail("No output file was expected with pattern" + pattern);
+    }
+  }
+
+  protected void failIfOutputFilePathContainingPatternDoesNotExists(FileSystem s3Client, Path targetDirPath, String pattern) throws IOException {
```
Comment (ebogi) [inline src/test/org/apache/sqoop/cloud/CloudImportJobTestCase.java]: Renaming of s3Client parameter has been missed. I would rename it to cloudFileSystem for example.

**Review by ebogi — CHANGES_REQUESTED** [review]
Hi Szabolcs, thank you for this improvement, it is very elegant! I ran S3 tests with your patch successfully. I had only a minor finding otherwise your change LGTM.

**Review by ebogi — APPROVED** [review]
Ship it!

**Comment by fszabo2** [comment]
Lgtm as well! Ship it!

**Pull request merged**